pkg/container-disk/container-disk.go

```diff
@@ -176,66 +213,113 @@ func GenerateContainers(vmi *v1.VirtualMachineInstance, podVolumeName string, bi
 	return generateContainersHelper(vmi, podVolumeName, binVolumeName, false)
 }
 
+func GenerateKernelBootContainer(vmi *v1.VirtualMachineInstance, podVolumeName string, binVolumeName string) *kubev1.Container {
+	if !util.IsKernelBootDefinedProperly(vmi) {
+		if f := vmi.Spec.Domain.Firmware; f != nil && f.KernelBoot != nil && f.KernelBoot.Container != nil &&
+			f.KernelBoot.Container.InitrdPath == "" && f.KernelBoot.Container.KernelPath == "" {
+			log.Log.Object(vmi).Warning("Kernel boot container does not contain kernel path nor initrd path - ignoring kernel-boot configuration")
+		}
+		return nil
+	}
```

Comment:
From my point of view, this validation should belong in the validation webhook[1] and we should be rejecting incorrect specs, instead of ignoring them. 
But I wonder what is your take on this?

[1] pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

Reply:
I need to fully understand something before answering that. Take a look on [this line of code](https://github.com/kubevirt/kubevirt/blob/1c9285807dac4684a3d3b0d77c0b94bc04a30609/pkg/virt-launcher/virtwrap/converter/converter.go#L1163). I'm converting the command line arguments to libvirt domain even if kernel/initrd images are not provided. From Roman's discussion on the previous PR I had the impression that this is the right way to go but now I'm re-thinking it.

Do command line arguments have effect even if no kernel/initrd images are provided? If there is an effect than a spec with only command line should be valid and command line args will be converted into the domain (sine `util.IsKernelBootDefinedProperly()` will return true). The validation you mentioned happens only in a single place before generating the container-disk (so that if kernel/initrd images aren't defined no containers will be generated).

If this has no effect though then I agree - the spec should be considered invalid and we can enforce that with validation webhook.

P.S. is it valid to provide initrd image only (without a kernel image)?

Reply:
Sorry, I shouldn't do reviews so late ;)


> I need to fully understand something before answering that. Take a look on [this line of code](https://github.com/kubevirt/kubevirt/blob/1c9285807dac4684a3d3b0d77c0b94bc04a30609/pkg/virt-launcher/virtwrap/converter/converter.go#L1163). I'm converting the command line arguments to libvirt domain even if kernel/initrd images are not provided. From Roman's discussion on the previous PR I had the impression that this is the right way to go but now I'm re-thinking it.

It is, afaik, I don't see any limitation here. 

> Do command line arguments have effect even if no kernel/initrd images are provided? If there is an effect than a spec with only command line should be valid and command line args will be converted into the domain (sine `util.IsKernelBootDefinedProperly()` will return true). The validation you mentioned happens only in a single place before generating the container-disk (so that if kernel/initrd images aren't defined no containers will be generated).

What I meant was that we shouldn't accept specs that are malformed...
For example
```
spec:
    firmware:
      kernelBoot:
        container:
          image: myimage/ext-kernel-boot-demo
```
Would we know where to take the kernel and initrd from this image?

If I read the code correctly, then you simply suggest to skip this container image, while I think we should explicitly reject this.


> P.S. is it valid to provide initrd image only (without a kernel image)?

afaik, yes.

Reply:
I would also rename `IsKernelBootDefinedProperly` to something like `HasKernelBootContainerImage` or `ShouldUseKernelBootContainerImage` or differently, just in a way that it will indicate whether we are going to use this container disk or not.

Reply:
I see your point. What went through my head is that someone might use the following configuration:

```
spec:
    firmware:
      kernelBoot:
        commandLine: "something valid"   
        container:
          image: myimage/ext-kernel-boot-demo
```

This definition is, as you claimed, invalid as an image is defined without a path to kernel/initrd images. But this definition is also partially valid in the sense that command line arguments can still be used.

As I see it in this situation we have 2 options - either to fail with an error or use the partially valid configuration (command-line args) and trigger a warning about the problematic definition which will be ignored. I thought the the later option is better then the first.

What do you think?

Reply:
And I agree, `IsKernelBootDefinedProperly` bugs me too. After we finish this discussion I will choose a better name :)

Reply:
> I see your point. What went through my head is that someone might use the following configuration:
> 
> ```
> spec:
>     firmware:
>       kernelBoot:
>         commandLine: "something valid"   
>         container:
>           image: myimage/ext-kernel-boot-demo
> ```
> 
> This definition is, as you claimed, invalid as an image is defined without a path to kernel/initrd images. But this definition is also partially valid in the sense that command line arguments can still be used.
> 
> As I see it in this situation we have 2 options - either to fail with an error or use the partially valid configuration (command-line args) and trigger a warning about the problematic definition which will be ignored. I thought the the later option is better then the first.
> 
> What do you think?

My take on this is that the validating webhook allows us to be explicit.
Rejecting such specs early allows us to reduce the amount of "guessing" we have to do to understand the original intent of the user.
The user then can quickly fix the issue and post an updated spec.

Alternatively, If we will just ignore a malformed container section, while the user expects to boot from this container image - he will spend a lot of time going through all the logs looking for what went wrong...
I guess I would prefer to avoid this and notify the user that something is wrong ahead of time.